### PR TITLE
storage/engine: Adjust SEEK_BOUND

### PR DESCRIPTION
**Pull request by MyonKeminta**
Currently in the implementation of `near_seek` and `near_seek_for_prev`, it first tries to move the iterator towards the target key. If the iterator doesn't reach the destination in 30 steps, call seek instead. However, a frequently-updating item in TiKV may have a large number of different versions, so near seek is more likely to fallback to `seek`. In this situation, too large `SEEK_BOUND` may reduce the performance.
A previously-done test shows that the time costs of `Seek`, `Next` and `Prev` operation of RocksDB's iterator approximately have such relationships when there are 100,000,000 kv pairs in the db:
Seek / Next ≈ 40
Seek / Prev ≈ 30
I think set SEEK_BOUND to 4 to 8 is OK.
Further more, costs of `Next` and `Prev` are not the same. Maybe it's better split `SEEK_BOUND` into two.

### CODE REVIEW

**Comment by siddontang** [comment]
> A previously-done test shows 

Where is the test?

File: src/storage/engine/mod.rs [hunk 1]
```diff
@@ -155,11 +156,11 @@ pub trait Iterator {
 }
 
 macro_rules! near_loop {
-    ($cond:expr, $fallback:expr, $st:expr) => ({
+    ($cond:expr, $fallback:expr, $bond:expr, $st:expr) => ({
```
Comment (zhangjinpeng1987) [inline src/storage/engine/mod.rs]: s/bond/bound

**Comment by huachaohuang** [comment]
Can you share some numbers about the effect?

**Comment by siddontang** [comment]
any update?

**Comment by MyonKeminta** [comment]
Sorry.
The test I mentioned before is here:
```
Seek/Next/Prev Benchmark

IDC2 172.16.20.1
CPU: 40 * Intel(R) Xeon(R) CPU E5-2630 v4 @ 2.20GHz
Run db_bench in /data2
Generating DB:
./db_bench
    --key_size=30 \
    --value_size=128 \
    --bloom_bits=10 \
    --benchmarks=fillrandom \
    --memtablerep=skip_list \
    --compression_type=Snappy \
    --compression_ratio=0.5 \
    --num=100000000 \
    --db=$file_path \
    --use_existing_db=false

Benchmarking:
./db_bench
    --key_size=30 \
    --value_size=128 \
    --threads=1 \
    --bloom_bits=10 \
    --db=$file_path \
    --num=100000000 \
    --reads=$3 \
    --use_existing_db=true \
    --memtablerep=skip_list \
    --compression_type=Snappy \
    --compression_ratio=0.5 \
    --benchmarks=seekrandom \
    --seek_nexts=$2 \
    $rev

Results:
1 Seek + 1 Next: 62718 ops/sec (--reads=10000000 --seek_nexts=1)
1 Seek + 1000 Next: 2570 ops/sec (--reads=100000 --seek_nexts=1000)
1 Seek + 1 ReverseToBackward + 1 Prev: 47422 ops/sec (--reads=10000000 --seek_nexts=1 --reverse_iterator)
1 Seek + 1 ReverseToBackward + 1000 Prev: 1917 ops/sec (--reads=100000 --seek_nexts=1 --reverse_iterator)

Seek / Next ≈ 41.7
Seek / Prev ≈ 31.1
```

However later I found that reducing SEEK_BOUND from 30 to 8 doesn't improve as much as expected according to sysbench (As is shown in my comment in #2689 )

Since #2689 is closed, I think it's better not merge this first. Further digging is needed.

**Comment by siddontang** [comment]
IMO, 15% improvement is very huge, we can go on.

**Comment by siddontang** [comment]
PTAL @huachaohuang

**Comment by MyonKeminta** [comment]
This pr doesn't seems improves much as expected. 15% is what #2689 did.

**Comment by zhangjinpeng1987** [comment]
LGTM

**Comment by huachaohuang** [comment]
According to the comment in https://github.com/pingcap/tikv/pull/2689, this PR has no noticeable improvement, am I right?
So, should we keep investing the result or close this?

**Comment by siddontang** [comment]
should investigate the problem.

**Comment by zhangjinpeng1987** [comment]
@MyonKeminta Please do the prev/next/seek/seek_for_prev benchmark on a 100GB RocksDB.

**Comment by breeswish** [comment]
According to my simple micro tests, SEEK_BOUND = 30 is too large. SEEK_BOUND == 4 might be appropriate.

**Comment by breeswish** [comment]
We have changed seek bound to 8 in master now.

**Comment by breeswish** [comment]
Seek bound mostly affects the performance when there are a lot of versions. For smaller versions, there is no difference.

**Pull request closed without merging**